framework/src/modules/chain/dpos/delegates.js

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+const { hash } = require('@liskhq/lisk-cryptography');
+
+const shuffleActiveDelegateList = (round, list) => {
+	const seedSource = round.toString();
+	const delegateList = [...list];
+	let currentSeed = hash(seedSource, 'utf8');
+
+	for (let i = 0, delCount = delegateList.length; i < delCount; i++) {
+		for (let x = 0; x < 4 && i < delCount; i++, x++) {
+			const newIndex = currentSeed[x] % delCount;
+			const b = delegateList[newIndex];
+			delegateList[newIndex] = delegateList[i];
+			delegateList[i] = b;
+		}
+		currentSeed = hash(currentSeed);
+	}
+
+	return delegateList;
+};
+
+class Delegates {
+	constructor({ storage, activeDelegates, exceptions }) {
+		this.delegateListCache = {};
+		this.storage = storage;
+		this.activeDelegates = activeDelegates;
+		this.exceptions = exceptions;
+	}
+
+	async getRoundDelegates(round) {
+		const list = await this.generateActiveDelegateList(round);
+		return shuffleActiveDelegateList(round, list);
+	}
+
+	async getDelegatePublicKeysSortedByVote() {
+		const filters = { isDelegate: true };
+		const options = {
+			limit: this.activeDelegates,
+			sort: ['vote:desc', 'publicKey:asc'],
+		};
+		const accounts = await this.storage.entities.Account.get(filters, options);
+		return accounts.map(account => account.publicKey);
+	}
+
+	async generateActiveDelegateList(round) {
+		if (this.delegateListCache[round]) {
+			return this.delegateListCache[round];
+		}
+
+		let delegatePublicKeys = await this.storage.entities.RoundDelegates.get(
+			round
+		);
+
+		if (!delegatePublicKeys.length) {
+			delegatePublicKeys = await this.getDelegatePublicKeysSortedByVote();
```

Comment:
Keeping in mind that behavior of this function will change based on the protocol version, which will mean that `RoundDelegates` table will insert entries for publicKeys sorted by vote with the current round number in the earlier protocol version and current round number -2 in the next protocol version, do you think maybe it makes more sense to separate creating entry in RoundDelegates in a different function which will be called independently?

Reply:
I think when we need to get the delegate list from the round -2, we can just pass it to the `generateActiveDelegateList` function. `getDelegatePublicKeysSortedByVote` doesn't need to know which round we are in. It's just for returning the top voted 101 delegates. Also `return shuffleDelegateListForRound(round, list)` function makes sure that list is shuffled for that round.

So I think this is fine.